### PR TITLE
[Animal] Implement refreshed Animal nutritional supply in `NutritionSupplyCalculator`

### PR DESCRIPTION
**Pull request by Isgotkowitz**
Implements routines which calculate the energy and nutrients supplied in ration.

### Context

Issue(s) closed by this pull request: addresses #2130 

### What

A new module is added, and is dedicated to calculating energy and nutrients supplied in a ration. The new module utilizes some of the data structures added in #2136 to gather the supplied amounts. 

### Why

Calculating the energy and nutrient contents of a ration is critical, because it the contents can then be compared to the energy and nutrient requirements of an animal. And this comparison indicates whether the ration satisfies the animal's requirements or not.

### How

The `nutrients` directory has the new class `NutritionSupplyCalculator`, which is just a collection of `@classmethod`s. `NutritionSupplyCalculator` takes in a ration as a mapping of RuFaS Feed IDs to amounts in kg. It also takes in all the nutrient information of the Feeds used in the form of `Feed` instances. The amounts of energy and nutrients calculated to be in the ration are put into `NutritionSupply` instance.

These equations were pulled from the constraint routines in `ration_driver.py`.

### Test plan

A full unit test suite has been written for `NutritionSupplyCalculator`.

### CODE REVIEW

File: tests/animal_module_tests/nutrient_tests/test_nutrition_supply_calculator.py [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+from RUFAS.biophysical.animal.data_types.nutrition_data_structures import NutritionSupply
+from RUFAS.biophysical.animal.nutrients.nutrition_supply_calculator import FeedInRation, NutritionSupplyCalculator
+from RUFAS.data_structures.feed_storage_animal_connection import Feed, RUFAS_ID, Type
+
+
+@pytest.fixture
+def feeds(mocker: MockerFixture) -> tuple[Feed, Feed, Feed]:
+    """Mock feeds to be used for testing."""
+    mocker.patch.object(Feed, "__init__", return_value=None)
+    feed_1, feed_2, feed_3 = Feed(), Feed(), Feed()
```
Comment (Isgotkowitz) [inline tests/animal_module_tests/nutrient_tests/test_nutrition_supply_calculator.py:13]: Mypy doesn't like this, but I think that is ok because this is a very clean, efficient way to make mock `Feed`s

File: tests/animal_module_tests/nutrient_tests/test_nutrition_supply_calculator.py [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+from pytest_mock import MockerFixture
+
+from RUFAS.biophysical.animal.data_types.nutrition_data_structures import NutritionSupply
+from RUFAS.biophysical.animal.nutrients.nutrition_supply_calculator import FeedInRation, NutritionSupplyCalculator
+from RUFAS.data_structures.feed_storage_animal_connection import Feed, RUFAS_ID, Type
+
+
+@pytest.fixture
+def feeds(mocker: MockerFixture) -> tuple[Feed, Feed, Feed]:
+    """Mock feeds to be used for testing."""
```
Comment (Isgotkowitz) [inline tests/animal_module_tests/nutrient_tests/test_nutrition_supply_calculator.py]:
```suggestion
    """Mock feeds used for testing."""
```

**Comment by Isgotkowitz** [comment]
@tomhuhh could you please check the methods `_calculate_calcium_supply` and `_calculate_phosphorus_supply` and their respective unit tests to see if the units are are being handled correctly?

**Comment by tomhuhh** [comment]
Hi @Isgotkowitz, I think another to test it is to assume there is only calcium in the ingredient, then it should return the amount of calcium fed * digestibility.
In order to do that, we can assume calcium content 100%

**Comment by tomhuhh** [comment]
Hello Ed! I also agree that if the unit of dry matter intake (`feed.amount`) is in kilograms, the unit of calcium and phosphorus supplies should also be in kilograms. I took a look at the original code under Routine->ration_driver.py. The code there performs correctly and it seems to me that they did not do a conversion of grams to kilograms there for the final outcome of calcium and phosphorus. Could there be a typo in the design doc instead?

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+            [
+                feed.amount
+                * calcium_digestibility[feed.info.rufas_id]
+                * feed.info.calcium
+                * GeneralConstants.PERCENTAGE_TO_FRACTION
+                for feed in feeds
+            ]
+        )
+
+        return total * GeneralConstants.KG_TO_GRAMS
```
Comment (Isgotkowitz) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py:340]: Note that this method (and the one for phosphorus below) should have their units scrutinized carefully. After discussion with @tomhuhh we think that we have them tested correctly, but I don't think that it is possible to get too many eyes on the results.

**Comment by Isgotkowitz** [comment]
> Hi @Isgotkowitz, I think another to test it is to assume there is only calcium in the ingredient, then it should return the amount of calcium fed * digestibility. In order to do that, we can assume calcium content 100%

Test cases added for both calcium and phosphorus!

> Hello Ed! I also agree that if the unit of dry matter intake (feed.amount) is in kilograms, the unit of calcium and phosphorus supplies should also be in kilograms. I took a look at the original code under Routine->ration_driver.py. The code there performs correctly and it seems to me that they did not do a conversion of grams to kilograms there for the final outcome of calcium and phosphorus. Could there be a typo in the design doc instead?

I think it is definitely possible. With the magic numbers in `calcium_constraint()` in `ration_driver.py`, I think it is absolutely possible that the units got mixed up.

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 4]
```diff
@@ -0,0 +1,651 @@
...
+        for feed in feeds:
+            if feed.info.feed_type is FeedComponentType.MINERAL:
+                energy = 0.0
+            elif feed.info.is_fat is True:
+                energy = feed.info.DE
+            elif feed.info.EE >= 3.0:
+                energy = 1.01 * actual_digestible_energy[feed.info.rufas_id] - 0.45 * 0.0046 * (feed.info.EE - 3.0)
+            else:
+                energy = 1.01 * actual_digestible_energy[feed.info.rufas_id] - 0.45
+            actual_metabolizable_energy[feed.info.rufas_id] = energy
```
Comment (Isgotkowitz) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py:159]: Ed: right now the actual metabolizable energy is in Mcal / kg. Check whether to see if this is correct or if this should be changed so that it is the total Mcal from each feed type.

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+            phosphorus=phosphorus,
+            dry_matter=dry_matter_intake,
+            ndf_content=ndf_content,
+            fat_content=fat_content,
+        )
+
+    @classmethod
+    def _calculate_discount(cls, feeds: list[FeedInRation], body_weight: float) -> float:
+        """
+        Calculates discount applied to Total Digestible Nutrients (TDN) and Digestible Energy (DE).
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py:83]: Is there a modifier to explain what the discount is/why it's necessary? Nutritional discount or storage degredation discount or financial discount?

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 6]
```diff
@@ -0,0 +1,651 @@
...
+            actual_metabolizable_energy[feed.info.rufas_id] = energy
+
+        return actual_metabolizable_energy
+
+    @classmethod
+    def _calculate_actual_maintenance_net_energy(
+        cls, feeds: list[FeedInRation], actual_metabolizable_energy: dict[RUFAS_ID, float]
+    ) -> float:
+        """
+        Calculates the actual net energy for maintenance of feeds.
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: Can you reword this? It's a little confusing what this is calculating. Is it the actual net energy from the feed needed for maintenance?

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 7]
```diff
@@ -0,0 +1,651 @@
...
+
+    @classmethod
+    def _calculate_actual_lactation_net_energy(
+        cls,
+        feeds: list[FeedInRation],
+        actual_metabolizable_energy: dict[RUFAS_ID, float],
+        actual_digestible_energy: dict[RUFAS_ID, float],
+    ) -> float:
+        """
+        Calculates the actual net energy for lactation of feeds.
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: Similarly if this can be reworded for clarity I think that would be helpful.

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 8]
```diff
@@ -0,0 +1,651 @@
...
+        total = sum([feed.amount * actual_lactation_net_energy[feed.info.rufas_id] for feed in feeds])
+
+        return total
+
+    @classmethod
+    def _calculate_actual_growth_net_energy(
+        cls, feeds: list[FeedInRation], actual_metabolizable_energy: dict[RUFAS_ID, float]
+    ) -> float:
+        """
+        Calculates actual net energy for growth of feeds.
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: Similar comment as for the others.

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 9]
```diff
@@ -0,0 +1,651 @@
...
+        References
+        ----------
+        Animal Scientific Documentation [A.Cow.E.16]-[A.Heifer.E.16]
+
+        """
+        calcium_digestibility: dict[RUFAS_ID, float] = {}
+
+        for feed in feeds:
+            if feed.info.feed_type is FeedComponentType.FORAGE:
+                ca_digestibility = 0.3
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]:
```suggestion
                digestibility = 0.3
```
minor suggestion to match the energy ones which were more general within the for loops.

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 10]
```diff
@@ -0,0 +1,651 @@
...
+                for feed in feeds
+            ]
+        )
+
+        return total * GeneralConstants.KG_TO_GRAMS
+
+    @classmethod
+    def _calculate_phosphorus_supply(cls, feeds: list[FeedInRation]) -> float:
+        """
+        Calculates the phosphorus supply in the ration (kg).
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]:
```suggestion
        Calculates the phosphorus supply in the ration.
```

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 11]
```diff
@@ -0,0 +1,651 @@
...
+    @classmethod
+    def _calculate_metabolizable_protein_supply(
+        cls,
+        feeds: list[FeedInRation],
+        dry_matter_intake: float,
+        actual_tdn_percentages: dict[RUFAS_ID, float],
+        body_weight: float,
+    ) -> float:
+        """
+        Calculates amount of metabolizable protein in ration (kg).
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]:
```suggestion
        Calculates amount of metabolizable protein in ration.
```
Not sure about adding this to the function description. I think if you're concerned about noting that the inputs are in kg and the output is in g, it should stay in the params and returns section (just my opinion).

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 12]
```diff
@@ -0,0 +1,651 @@
...
+        ----------
+        feeds : list[FeedInRation]
+            List of feeds in ration, including the amount and nutritive properties.
+        dry_matter_intake : float
+            Total dry matter contained in the ration (kg).
+
+        Returns
+        -------
+        float
+            Percentage of the ration's dry matter which is made up of concentrates.
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]:
```suggestion
            Percentage of the ration's dry matter which is made up of concentrates (percent).
```

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 13]
```diff
@@ -0,0 +1,651 @@
...
+            Total dry matter contained in the ration (kg).
+        body_weight : float
+            The body weight of an animal (kg).
+        percentage_concetrates : dict[RUFAS_ID, float]
+            Percentage of the ration's dry matter which is made up of concentrates.
+
+        Returns
+        -------
+        dict[RUFAS_ID, float]
+            Mapping of RuFaS Feed IDs to protein passage rates (percentage / hour).
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py:515]: Likely need to add this unit to units.py

**Review by ew3361zh — COMMENTED** [review]
Looking good! I am not able to run either the simulation nor the test suite on this branch though. Errors below:

from running the simulation:
Screenshot 2024-12-23 at 4 11 28 PM

from running the test suite:
Screenshot 2024-12-23 at 4 11 05 PM

**Comment by Isgotkowitz** [comment]
> Looking good! I am not able to run either the simulation nor the test suite on this branch though. Errors below:
> 
> from running the simulation: Screenshot 2024-12-23 at 4 11 28 PM
> 
> from running the test suite: Screenshot 2024-12-23 at 4 11 05 PM

Thank you for the review @ew3361zh. Unfortunately this PR is not at a point yet where the full simulation or test suite can be run, it is stacked on top of other PRs which are works in progress.

`test_nutrition_supply_calculator.py` should be fully passing when run on its own though.

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 5 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py:83]: Great question. It is some sort of nutritional discount, because it gets applied to TDN and digestible energy. I'm sure it has some sort of specific/specialized named though, @KFosterReed and @tomhuhh do either of you have suggestions for how we could name this discount in a clearer way?

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 4 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py:159]: This is correct. The energy content / kg is used later on down the line to calculate other totals.

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 6 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: I believe it is the actual net energy from the feeds available to be used for maintenance. I'll update it.

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 10 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: Good catch, the units should not be there and kg is incorrect anyways

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 13 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py:515]: I think we should wait on that until there is a RuFaS output that uses this unit.

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 12 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: I think in this case it would be a little redundant to put "(percent)" at the end of the description, because the description makes is already very explicit about what the unit is

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 5 above)
Comment (KFosterReed) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py:83]: Hmm, I don't think there is a super obvious name that comes to mind. The discount reduces the amount of the nutrients that are available as the intake level gets higher. Kinda the same idea as if water is moving through a pipe with a small hole in it, the amount of water that will leak out per unit of total water will decrease as the flow rate increases. 

Anyway, a couple of options for improving the clarity could be: 

"intake_discount" or "intake_nutrient_discount"

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 14]
```diff
@@ -0,0 +1,651 @@
...
+        """
+        actual_metabolizable_energy: dict[RUFAS_ID, float] = {}
+
+        for feed in feeds:
+            if feed.info.feed_type is FeedComponentType.MINERAL:
+                energy = 0.0
+            elif feed.info.is_fat is True:
+                energy = feed.info.DE
+            elif feed.info.EE >= 3.0:
+                energy = 1.01 * actual_digestible_energy[feed.info.rufas_id] - 0.45 * 0.0046 * (feed.info.EE - 3.0)
```
Comment (KFosterReed) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: Small typo in equation. Adjustment for fat content should be added to the original equation instead of multiplied by 0.45
`energy = 1.01 * actual_digestible_energy[feed.info.rufas_id] - 0.45  +  0.0046 * (feed.info.EE - 3.0)`

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 15]
```diff
@@ -0,0 +1,651 @@
...
+            ]
+        )
+        ration_rdp_content = sum(
+            [
+                feed.amount * rdp_percentages[feed.info.rufas_id] * GeneralConstants.PERCENTAGE_TO_FRACTION
+                for feed in feeds
+            ]
+        )
+
+        metabolizable_protein_tdn = ration_tdn_content * 0.13 * GeneralConstants.KG_TO_GRAMS
```
Comment (KFosterReed) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: Suggest change "metabolizable_protein_tdn" to "microbial_protein_tdn" and

"metabolizable_protein_rdp" to "microbial_protein_rdp".

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 16]
```diff
@@ -0,0 +1,651 @@
...
+        metabolizable_protein_rdp = ration_rdp_content * 0.85 * GeneralConstants.KG_TO_GRAMS
+        metabolizable_bacterial_protein_production = float(
```
Comment (KFosterReed) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: I know this is the name that is in the documentation but I am going to change it to be "metabolizable microbial protein" because it is more accurate. The protein is actually from a mixture of bacteria, protozoa and archaea so 'microbes' captures all of them where as bacteria implies it is just part of the microbial protein supply. 

TLDR please change to "metabolizable_microbial_protein_production"

**Review by KFosterReed — APPROVED** [review]
With a few minor variable name changes this looks good to me!

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 14 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: Nice catch! Would you like me to update this or make a note on it in the documentation?

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 16 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: Updated! Thanks for the explanation

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 5 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py:83]: Updated! Thanks for the suggestions Kristan

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 14 above)
Comment (KFosterReed) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: I'm updating our documentation but thanks for the offer!

**Comment by Isgotkowitz** [comment]
#2136 has been rebased off of `dev`, but this PR has not been rebased yet which is why it appears to have such a massive amount of changes

Edit: rebased!

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 12 above)
Comment (ew3361zh) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: My thought is that the format should follow suit. e.g. we could rewrite this (from the next function):

`Total dry matter contained in the ration (kg).`

as

`Total kg of dry matter contained in the ration.`

But the standard we've set is to have it be the first way.
I understand the redundancy concern but I think my preference would be for it to match the style with the unit at the end (could be rewritten as `Portion of the ration's dry matter which is made up of concentrates (percent).`)

**Review by ew3361zh — APPROVED** [review]
Looks good! Nice work organizing all the code into a very readable format. Building a nice foundation.

**Comment by Isgotkowitz** [comment]
> Looks good! Nice work organizing all the code into a very readable format. Building a nice foundation.

Thanks Niko! I appreciate all the reviews!

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py (hunk 12 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py]: Ok, sounds good!

File: RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py [hunk 17]
```diff
@@ -0,0 +1,649 @@
...
+            Maps the RuFaS ID of a feed to the amount fed in a ration (kg dry matter).
+
+        Returns
+        -------
+        NutritionSupply
+            NutritionSupply instance containing the energy and nutritive content of the given ration.
+
+        """
+        feeds = [
+            FeedInRation(amount=amount, info=next((feed for feed in feeds_used if feed.rufas_id == rufas_id), None))
```
Comment (Isgotkowitz) [inline RUFAS/biophysical/animal/nutrients/nutrition_supply_calculator.py:42]: Mypy doesn't like this, because there is the possibility that no matching feed information will be found and `None` will be used instead. This is a deliberate choice though, if a matching feed is not found that indicates a serious error, and will result in a NoneType error which will halt the simulation.

File: changelog.md [hunk 18]
```diff
@@ -74,7 +74,8 @@ v0.9.2
 - [2149](https://github.com/RuminantFarmSystems/MASM/pull/2149) - [minor change] [Crop and Soil] Relocates TillageImplements to data structures dir.
 - [2137](https://github.com/RuminantFarmSystems/MASM/pull/2137) - [minor change] [Feed] Make post_wilting_moisture_percentage and user input. 
 - [2134](https://github.com/RuminantFarmSystems/MASM/pull/2134) - [minor change] [Crop & Soil] Adds ammonium fraction to fertilizer mixes input.
-- [2136](https://github.com/RuminantFarmSystems/MASM/pull/2137) - [minor change] [Animal] Adds data structures to support calculation of animal nutrient requirements and supply.
+- [2136](https://github.com/RuminantFarmSystems/MASM/pull/2136) - [minor change] [Animal] Adds data structures to support calculation of animal nutrient requirements and supply.
+- [2138](https://github.com/RuminantFarmSystems/MASM/pull/2136) - [minor change] [Animal] Refreshes module which calculates the nutrients supplied to an animal by a ration.
```
Comment (Isgotkowitz) [inline changelog.md]:
```suggestion
- [2138](https://github.com/RuminantFarmSystems/MASM/pull/2138) - [minor change] [Animal] Refreshes module which calculates the nutrients supplied to an animal by a ration.
```
